File: source/com/barrybecker4/puzzle/tantrix1/solver/path/permuting/SameTypeTileMixer.java
// Copyright by Barry G. Becker, 2012. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.puzzle.tantrix1.solver.path.permuting;

import com.barrybecker4.common.math.MathUtil;
import com.barrybecker4.puzzle.tantrix1.solver.path.PathType;
import com.barrybecker4.puzzle.tantrix1.solver.path.TantrixPath;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

/**
 * Mix tiles of the same type without changing the primary originalPath.
 *
 * @author Barry Becker
 */
public class SameTypeTileMixer {

    private static final int MAX_ITER = 20;
    private PathType type;
    private TantrixPath originalPath;

    /**
     * Constructor
     * @param type one of the 3 path types
     * @param path the path to create permutations of.
     */
    public SameTypeTileMixer(PathType type, TantrixPath path) {
        this.type = type;
        this.originalPath = path;
    }

    public List<TantrixPath> findPermutedPaths() {

        TilesOfTypeIndices indices = new TilesOfTypeIndices(type, originalPath);
        return findPermutedPaths(indices);
    }

    private List<TantrixPath> findPermutedPaths(TilesOfTypeIndices indices) {

        List<TantrixPath> permutedPaths = new LinkedList<>();
        PathTilePermuter permuter = new PathTilePermuter(originalPath);

        if (indices.size() == 2) {
            List<Integer> newOrder = new ArrayList<>(indices);
            Collections.reverse(newOrder);
            TantrixPath permutedPath = permuter.permute(newOrder, indices);
            permutedPaths.add(permutedPath);
        }
        else if (indices.size() > 2) {
            // add the original originalPath for now, to be sure we do not duplicated it, but return at end.
            permutedPaths.add(originalPath);

            int numIter = Math.min(originalPath.size() + 1, MAX_ITER);

            for (int i = 0; i < numIter; i++) {
                List<Integer> newOrder = new ArrayList<>(indices);
                Collections.shuffle(newOrder, MathUtil.RANDOM);
                TantrixPath permutedPath = permuter.permute(newOrder, indices);
                if (!permutedPaths.contains(permutedPath)) {
                    permutedPaths.add(permutedPath);
                }
            }
            permutedPaths.remove(0);
        }
        return permutedPaths;
    }
}

File: scala-source/com/barrybecker4/puzzle/tantrix/solver/path/permuting/SameTypeTileMixer.scala
// Copyright by Barry G. Becker, 2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.puzzle.tantrix.solver.path.permuting

import com.barrybecker4.puzzle.tantrix.solver.path.PathType
import com.barrybecker4.puzzle.tantrix.solver.path.TantrixPath

import scala.collection.mutable.ListBuffer
import scala.util.Random


object SameTypeTileMixer {
  private val MAX_ITER = 20
}

/**
  * Mix tiles of the same type without changing the primary originalPath.
  * @param pathType one of the 3 path types
  * @param originalPath the path to create permutations of.
  * @author Barry Becker
  */
class SameTypeTileMixer(var pathType: PathType, var originalPath: TantrixPath, rnd: Random = new Random()) {

  def findPermutedPaths: ListBuffer[TantrixPath] = {
    val indices = new TilesOfTypeIndices(pathType, originalPath)
    findPermutedPaths(indices)
  }

  private def findPermutedPaths(indices: TilesOfTypeIndices) = {
    val permutedPaths = new ListBuffer[TantrixPath]
    val permuter = new PathTilePermuter(originalPath)
    if (indices.size == 2) {
      val newOrder: ListBuffer[Int] = indices.list.reverse
      val permutedPath = permuter.permute(newOrder, indices.list)
      permutedPaths.append(permutedPath)
    }
    else if (indices.size > 2) {
      // add the original originalPath for now, to be sure we do not duplicate it, but remove it before returning.
      permutedPaths.append(originalPath)
      val numIter = Math.min(originalPath.size + 1, SameTypeTileMixer.MAX_ITER)

      for (i <- 0 until numIter) {
        val newOrder = rnd.shuffle(indices.list)  // permutations.next()

        //println("valOrder = " + newOrder)
        val permutedPath = permuter.permute(indices.list, newOrder)
        if (!permutedPaths.contains(permutedPath))
          permutedPaths.append(permutedPath)
      }
      permutedPaths.remove(0)
    }
    //println("The number of permuted paths = " + permutedPaths.size)
    permutedPaths
  }
}
